intl_translation_format/lib/src/utils/icu_message.dart

```diff
@@ -0,0 +1,166 @@
+      (genderKeyword & openCurly & interiorText & closeCurly)
+          .trim()
+          .map((result) => [result[0], result[2]]);
+  Parser get gender =>
+      preface & selectLiteral & comma & genderClause.plus() & closeCurly;
+  Parser get intlGender => gender.map((values) {
+       _extractArgumentIfNeeded(values.first);
+        Gender.from(values.first, values[3], null);
+      });
+  Parser get selectClause =>
+      (id & openCurly & interiorText & closeCurly).map((x) => [x.first, x[2]]);
+  Parser get generalSelect =>
+      preface & selectLiteral & comma & selectClause.plus() & closeCurly;
+  Parser get intlSelect => generalSelect.map(
+        (values) {
+          _extractArgumentIfNeeded(values.first);
+          return Select.from(values.first, values[3], null);
+        },
+      );
+
+  Parser get pluralOrGenderOrSelect => intlPlural | intlGender | intlSelect;
+
+  Parser get contents => pluralOrGenderOrSelect | parameter | messageText;
+  Parser get simpleText => (nonIcuMessageText | parameter | openCurly).plus();
+  Parser get empty => epsilon().map((_) => '');
+
+  Parser get parameter => (openCurly & id & closeCurly).map((param) {
+        _extractArgumentIfNeeded(param[1]);
+        return VariableSubstitution.named(param[1], null);
+      });
+
+  /// The primary entry point for parsing. Accepts a string and produces
+  /// a parsed representation of it as a Message.
+  Parser get message => (pluralOrGenderOrSelect | simpleText | empty)
+      .map((chunk) => Message.from(chunk, mainMessage));
+
+  final bool extractArguments;
+  final MainMessage mainMessage;
+
+  _IcuParser([this.mainMessage, this.extractArguments = false]) {
+    // There is a cycle here, so we need the explicit set to avoid
+    // infinite recursion.
+    interiorText.set(contents.plus() | empty);
+  }
+
+  // If extractArguments is true and a MainMessage parent exits it will 
+  // extract the arguments to the MainMessage
+  // 
+  // This is used when the templated is not generated from intl dart code.
+  _extractArgumentIfNeeded(String arg) {
+    if (mainMessage == null || !extractArguments) return;
+    mainMessage.arguments ??= [];
+    if (!mainMessage.arguments.contains(arg)) {
+      mainMessage.arguments.add(arg);
+    }
+  }
+}
+
+class IcuMessage {
+  static Message fromIcu(String icuString, [Message parent]) {
+    final _IcuParser parser = _IcuParser();
+    final Message message = parser.message.parse(icuString).value;
+    return message..parent = parent;
+  }
+}
+
+/// A main message generated from a icu string.
+/// The arguments from MainMessage are extracted from
+/// the message instead from the dart functions
+class IcuMainMessage extends MainMessage {
+  IcuMainMessage._();
+
+  factory IcuMainMessage(String icuString) {
+    final message = IcuMainMessage._();
+
+    final _IcuParser parser = _IcuParser(message, true);
+    message.addPieces([parser.message.parse(icuString).value]);
+
+    return message;
+  }
+}
```

Comment:
I think this would be a nice way to allow to create MainMessages without need them to be extracted from intl dart code. 

This _IcuParser adds the arguments to the MainMessage during parsing (only if extractArguments == true).

In formats not using ICU, it would have to be added manually during parsing too.

This would allow generating the messages dart code from an arb without extraction needed (we would still have to create the intl classes), but for now it improves also the testing.

Reply:
Yes, I like this.